scylla-cdc/src/stream_generations.rs

```diff
@@ -0,0 +1,540 @@
+use futures::stream::StreamExt;
+use scylla::batch::Consistency;
+use scylla::frame::response::result::Row;
+use scylla::frame::value::Timestamp;
+use scylla::query::Query;
+use scylla::{IntoTypedRows, Session};
+use std::error::Error;
+use std::sync::Arc;
+use uuid::Uuid;
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Ord, PartialOrd)]
+pub struct Generation(chrono::Duration);
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Ord, PartialOrd)]
+pub struct StreamID(Vec<u8>);
+
+/// Component responsible for managing stream generations.
+pub struct GenerationFetcher {
+    generations_table_name: String,
+    streams_table_name: String,
+    session: Arc<Session>,
+}
+
+// Number taken from: https://www.scylladb.com/2017/11/17/7-rules-planning-queries-maximum-performance/.
+const DEFAULT_PAGE_SIZE: i32 = 5000;
+
+impl GenerationFetcher {
+    /// Creates new instance of GenerationFetcher.
+    pub fn new(session: &Arc<Session>) -> GenerationFetcher {
+        GenerationFetcher {
+            generations_table_name: "system_distributed.cdc_generation_timestamps".to_string(),
+            streams_table_name: "system_distributed.cdc_streams_descriptions_v2".to_string(),
+            session: Arc::clone(session),
+        }
+    }
+
+    // Function instead of constant for testing purposes.
+    fn get_all_stream_generations_query(&self) -> String {
+        format!(
+            r#"
+    SELECT time
+    FROM {}
+    WHERE key = 'timestamps';
+    "#,
+            self.generations_table_name
+        )
+    }
+
+    /// In case of a success returns a vector containing all the generations in the database.
+    /// Propagates all errors.
+    pub async fn fetch_all_generations(&self) -> Result<Vec<Generation>, Box<dyn Error>> {
+        let mut query = Query::new(self.get_all_stream_generations_query());
+        select_consistency(&self.session, &mut query).await?;
```

Comment:
How about a function

```
async fn new_distributed_system_query(stmt: String, session: &Session) -> Result<Query, Box<dyn Error>>
```

which implements this pattern (`Query::new` + `select_consistency`)? This `select_consistency` function is only used in such scenarios.